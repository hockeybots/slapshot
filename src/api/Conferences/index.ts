import { CONFERENCES_ENDPOINT } from '../../endpoints';
import { Conference } from '../types';

import Endpoint from '../Endpoint';

/**
 * Conferences endpoint wrapper
 * @extends Endpoint
 * @description An abstraction of the {CONFERENCE_ENDPOINT} with a fluent API
 */
class Conferences extends Endpoint {
  /**
   * This method will transform API data in to a Conference object.
   * @param apiData {object} - The conference object of the response from the NHL API
   * @returns {Conference}
   */
  public static async toConference(apiData: any): Promise<Conference> {
    const conference: Conference = {
      abbreviation: apiData.abbreviation,
      active: apiData.active,
      id: apiData.id,
      name: apiData.name,
      shortName: apiData.shortName,
    };
    return conference;
  }

  private conferenceIds: Array<number>;
  constructor(...ids: Array<number>) {
    super();
    this.conferenceIds = ids;
    this.initUri();
  }
  public clear() {
    this.initUri();
  }
  public initUri() {
    this.uri = `${CONFERENCES_ENDPOINT}?conferenceId=${this.conferenceIds.join(',')}`;
  }
  /**
   * @description This method is used after building of the URI is complete. It will fetch and parse the NHL API data.
   * @returns {Array<Conference>}
   */
  public async data(): Promise<Array<Conference>> {
    try {
      const apiData = await this.load();
      return this.parseData<Conference>(apiData, 'conferences', Conferences.toConference);
    } catch (error) {
      return Promise.reject(error);
    }
  }
}

export default Conferences;
