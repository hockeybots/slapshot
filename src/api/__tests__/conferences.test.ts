import Conferences from '../Conferences';

describe('Api', () => {
  describe('Conferences', () => {
    describe('uri', () => {
      describe('when requesting one conference', () => {
        const conference = new Conferences(1);

        it('should return the correct uri', () => {
          expect(conference.uri).toBe('https://statsapi.web.nhl.com/api/v1/conferences?conferenceId=1');
        });
      });
      describe('when requesting multiple conferences', () => {
        const conference = new Conferences(1, 2);

        it('should return the correct uri', () => {
          expect(conference.uri).toBe('https://statsapi.web.nhl.com/api/v1/conferences?conferenceId=1,2');
        });
      });
      describe('toConference', () => {
        const mockApiData = {
          abbreviation: 'WW',
          active: true,
          id: 1,
          name: 'Wild West',
          shortName: 'Wild',
        };
        it('should return a conference object with the correct data', async () => {
          const conference = await Conferences.toConference(mockApiData);
          expect(conference).toMatchSnapshot();
        });
      });
    });
  });
});
