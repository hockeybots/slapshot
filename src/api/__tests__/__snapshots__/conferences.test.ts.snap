// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Api Conferences uri toConference should return a conference object with the correct data 1`] = `
Object {
  "abbreviation": "WW",
  "active": true,
  "id": 1,
  "name": "Wild West",
  "shortName": "Wild",
}
`;
